docs/design/resources.md

```diff
@@ -0,0 +1,120 @@
+# Operator Resources
+
+The etcd cluster operator extends the Kubernetes API with two new
+resource definitions:
+
+* `EtcdCluster`, or the "cluster resource", defines an entire
+  operating cluster.
+* `EtcdPeer`, or the "peer resource", defines a single peer of the
+  etcd cluster. This principally concerns a Pod, but also includes
+  attached storage and network identity.
+
+## The Peer Resource
+
+The peer resource represents a single peer in an etcd
+cluster. Encoding its bootstrap instructions, persistence settings,
+and extra pod settings (e.g., overridden CPU limits). Its `status`
+field exposes information about the peer including its hostname. In
+future the status may also include the liveness status if it's
+currently the leader of the etcd cluster.
+
+When reconciling the etcd peer resource the operator will do the
+following:
+
+1. If a correctly named persistent volume claim (PVC) does not exist,
+   create one with the settings provided on the peer resource.
+
+   We never delete a persistent volume claim automatically, so this
+   behaviour ensures that the operator will resume stopped clusters if
+   the storage remains even if the peer resource was deleted. In
+   future, automatic cluster recovery tools **may** delete PVC
+   resources if they determine it is required and safe to do so,
+   although this is outside of the scope of this document.
+
+   If the storage options on the peer resource are different from an
+   existing PVC, the operator makes no attempt to 'correct' the
+   preexisting PVC.
+2. If a correctly named pod does not exist, an etcd pod will be
+   launched with the PVC from above mounted as a volume. The bootstrap
+   configuration will always be provided to the pod as configuration,
+   as etcd will ignore bootstrap configuration if a data directory
+   already exists.
+3. If the pod is in a ready state, the operator will attempt to
+   contact it using the gRPC endpoint and retrieve information to
+   populate the status field. Otherwise, it will report the peer as
+   not-ready on the status field.
+
+## The Cluster Resource
+
+The cluster resource provides management of the entire cluster. It
+will ensure that the correct number of `EtcdPeer` resources exist,
+deploys a "headless" service to allow connectivity to the cluster, and
```

Comment:
have we considered making service deployment the responsibility of the user, rather than have it created by the operator? as long as the cluster has some predictable way of selecting peers that shouldn't be a problem, and it feels more k8s native to me. we'd have to figure out another way of doing the health check on the cluster resource but there are ways to work around that (e.g. list all Pods/EtcdPeers with a selector and check the status of each)

Reply:
I'm not sure what the benefit of leaving service creation up to the user is, and it feels like it's more work and coordination between the user and the controller/operators behaviour. etcd needs a predictable hostname for each peer, and if this changes it can cause potentially catastrophic damage (i.e. unable to recover quorum as members aren't able to discover each other).

Potentially *allowing* users to skip service creation in future is a good idea (perhaps by specifying something like `etcdcluster.spec.headlessServiceName`, but IMO the 'happy path' should be handled for you, and it's easier for us to add this feature than it is to take away I think.

Reply:
so if i have the use case "i want to expose my cluster under a public IP" would the expectation be that I would create my own service with the configuration I need, rather than attempting to modify/configure the generated service to do what I want?

if yes - that's fine. i just initially read this as "we'll create a service so you don't have to" rather than "we create a service to help things work internally"

Reply:
I'd like to have a 'batteries included' experience for the user, thus having the operator make the service for you. But that's just so that if you create a `EtcdCluster` with name `foo-bar`, you can just dial `foo-bar` as a service and get to your cluster.

Completely fine with having an override flag to turn this off if you wanted to make your own service.

Reply:
> if yes - that's fine. i just initially read this as "we'll create a service so you don't have to" rather than "we create a service to help things work internally"

Correct, I'd imagine we create an internal, controlled service that meets our requirements. Users would then be responsible for creating their *own* Service if the in-cluster one does not meet requirements (e.g. needs exposing externally).

Reply:
So I take it from this that we're happy with the operator creating a `service` as written, with future configuration and possibly disabling?